src/Compilers/CSharp/Test/Semantic/Semantics/NameOfTests.cs

```diff
@@ -1488,5 +1488,355 @@ public void nameof(string x)
+                 expectedOutput: "Length").VerifyDiagnostics();
+        }
+
+        [Fact, WorkItem(40229, "https://github.com/dotnet/roslyn/issues/40229")]
+        public void TestCanReferenceInstanceMembersFromAttributeInNameof()
+        {
+            var source = @"
+public class C
+{
+    [System.Obsolete(nameof(S.Length))]
+    public int P { get; }
+    public string S { get; }
+}";
+            CreateCompilation(source, parseOptions: TestOptions.RegularPreview).VerifyDiagnostics();
+        }
+
+        [Fact, WorkItem(40229, "https://github.com/dotnet/roslyn/issues/40229")]
+        public void TestCanReferenceInstanceMembersFromConstructorInitializersInNameof()
+        {
+            var source = @"
+public class C
+{
+    public C(string s){}
+    public C() : this(nameof(S.Length)){}
+    public string S { get; }
+}";
+            CreateCompilation(source, parseOptions: TestOptions.RegularPreview).VerifyDiagnostics();
+        }
+
+        [Fact, WorkItem(40229, "https://github.com/dotnet/roslyn/issues/40229")]
+        public void TestCanAccessStructInstancePropertyInLambdaInNameof()
+        {
+            var source = @"
+using System;
+
+public struct S
+{
+    public string P { get; }
+    public void M(ref string x)
+    {
+        Func<string> func = () => nameof(P.Length);
+    }
+}";
+            CreateCompilation(source, parseOptions: TestOptions.RegularPreview).VerifyDiagnostics();
+        }
+
+        [Fact, WorkItem(40229, "https://github.com/dotnet/roslyn/issues/40229")]
+        public void TestCanReferenceStaticMembersFromInstanceMemberInNameof1()
+        {
+            var source = @"
+public class C
+{
+    public C Prop { get; }
+    public static int StaticProp { get; }
+    public string M() => nameof(Prop.StaticProp);
+}";
+            CreateCompilation(source, parseOptions: TestOptions.RegularPreview).VerifyDiagnostics();
+        }
+
+        [Fact, WorkItem(40229, "https://github.com/dotnet/roslyn/issues/40229")]
+        public void TestCanReferenceStaticMembersFromInstanceMemberInNameof2()
+        {
+            var source = @"
+public class C
+{
+    public C Prop { get; }
+    public static int StaticProp { get; }
+    public static string M() => nameof(Prop.StaticProp);
+}";
+            CreateCompilation(source, parseOptions: TestOptions.RegularPreview).VerifyDiagnostics();
+        }
+
+        [Fact, WorkItem(40229, "https://github.com/dotnet/roslyn/issues/40229")]
+        public void TestCanReferenceStaticMembersFromInstanceMemberInNameof3()
+        {
+            var source = @"
+public class C
+{
+    public C Prop { get; }
+    public static string M() => nameof(Prop.M);
+}";
+            CreateCompilation(source, parseOptions: TestOptions.RegularPreview).VerifyDiagnostics();
+        }
+
+        [Fact, WorkItem(40229, "https://github.com/dotnet/roslyn/issues/40229")]
+        public void TestCanReferenceStaticMembersFromInstanceMemberInNameof4()
```

Comment:
>TestCanReferenceStaticMembersFromInstanceMemberInNameof4 [](http://example.com/codeflow?start=20&length=56)

We should have at least one test that targets ```TestOptions.RegularNext``` for each distinct code path that eliminates errors.